### ISSUE TITLE
Wrong error code when denormalizing a non existant item

### ISSUE THREAD

**Issue by KDederichs** [api-platform/core#6116]
**API Platform version(s) affected**: 3.2.11

**Description**  
While writing tests I noticed that when you try to denormalize a non existing item it'll throw an exception with statuscode 500.

It shouldn't error out with a 500 error since that's not the expected error code for cases like that.

**How to reproduce**  
```php
#[ApiResource]
class Foo {
   prive ?Bar $bar = null;
}
#[ApiResource]
class Bar {
}
```
and then try to create a Foo with a reference to a non existing Bar.
**Possible Solution**  

There's two solutions for this:
Change the error code to 404, to match the not found paradime.

Optionally allow not found items to serialize to null so users can throw their own errors with Validations.

**Additional Context**  
```
    "title": "An error occurred",
    "detail": "Item not found for \"/api/bar/018d1ca6-da89-77c2-93a9-2b3593055e42\".",
    "status": 500,
    "type": "/errors/500",
    "trace": [
        {
            "file": "/app/vendor/api-platform/core/src/Serializer/AbstractItemNormalizer.php",
            "line": 860,
            "function": "denormalizeRelation",
            "class": "ApiPlatform\\Serializer\\AbstractItemNormalizer",
            "type": "->"
        },
...
```

**Comment by soyuka** [comment]
I'm pretty sure we've a test for this, imo it should even send a 400 (I have a deja-vu feeling)

**Comment by KDederichs** [comment]
Digging into it, it's running into this exception:
https://github.com/api-platform/core/blob/ac8031eb42810e6c538ffaca295cf7b4d35adf06/src/Serializer/AbstractItemNormalizer.php#L546

Maybe the test is the other one with the context?

**Comment by KDederichs** [comment]
Yeah I don't think that test covers it.
I just switched out the exception with a `return null` and the `relation.feature` test keeps running the same as before

**Comment by Skkay** [comment]
Same issue with a backed enum:

Example:
```php
<?php

enum StatusEnum: string {
    case Ready: 'ready';
}

class MyEntity {
    private ?StatusEnum $status = null;

    public function getStatus(): ?StatusEnum
    {
        return $this->status;
    }

    public function setStatus(StatusEnum $status): static
    {
        $this->status = $status;

        return $this;
    }
}
```

Making a POST request with:
```json
{
    "status": "unknown_status",
}
```

Will result in an Error 500:
```jsonc
{
    "@context": "\/api\/contexts\/Error",
    "@type": "hydra:Error",
    "hydra:title": "An error occurred",
    "hydra:description": "The data must belong to a backed enumeration of type App\\Enum\\StatusEnum",
    "trace": [
        // ...
    ]
}
```

**Comment by webda2l** [comment]
> Same issue with a backed enum:
> 
> Example:
> 
> ```
> <?php
> 
> enum StatusEnum: string {
>     case Ready: 'ready';
> }
> 
> class MyEntity {
>     private ?StatusEnum $status = null;
> 
>     public function getStatus(): ?StatusEnum
>     {
>         return $this->status;
>     }
> 
>     public function setStatus(StatusEnum $status): static
>     {
>         $this->status = $status;
> 
>         return $this;
>     }
> }
> ```
> 
> Making a POST request with:
> 
> ```json
> {
>     "status": "unknown_status",
> }
> ```
> 
> Will result in an Error 500:
> 
> ```js
> {
>     "@context": "\/api\/contexts\/Error",
>     "@type": "hydra:Error",
>     "hydra:title": "An error occurred",
>     "hydra:description": "The data must belong to a backed enumeration of type App\\Enum\\StatusEnum",
>     "trace": [
>         // ...
>     ]
> }
> ```

Same with latest 3.2.17

**Comment by soyuka** [comment]
I could not reproduce can you help by improving my test at https://github.com/soyuka/core/commit/2d1a5115a601127dacdfa6b896717e1909cea527 ?

**Comment by KDederichs** [comment]
@soyuka 

When I run your test and add 
```
    And the JSON should be a superset of:
    """
    {
      "@id": "/errors/400",
      "@type": "hydra:Error",
      "title": "An error occurred",
      "detail": "Item not found for \"/issue6116_relation/1\""
    }
    """
```

to the test it fails with 
```
      +  'detail' => 'Invalid IRI "/issue6116_relation/1".',
```


So looks like you get the right status code for the wrong reason

**Comment by soyuka** [comment]
Not sure I understand, 400 is the correct status code.

**Comment by KDederichs** [comment]
Sure, but the 500 happens when you get a "Item not found" error, not an "Invalid Iri" error.
So it would seem the test doesn't seem to test the actual problem but something else.

**Comment by KDederichs** [comment]
Like the test setup seems to be flawed in a way.
Looking into it more it throws Invalid Iri in `AbstractItemNormalizer.php:572` because no route matches `"/issue6116_relation/1"`. 
The route should exist but shouldn't find the resource.

What you got now seems to be just a non existing route for some reason

**Comment by KDederichs** [comment]
Hmm still works, I'm trying to track down what's causing it

**Comment by KDederichs** [comment]
Ok found the issue @soyuka 

The error happens if you define custom exception mappings:
```
    exception_to_status:
        App\Exception\OrganisationNotSetException: !php/const Symfony\Component\HttpFoundation\Response::HTTP_FORBIDDEN
        App\Exception\ProjectNotSetException: !php/const Symfony\Component\HttpFoundation\Response::HTTP_FORBIDDEN
```

It maybe overwrites the ones you guys do?

**Comment by KDederichs** [comment]
Also in the test bundle's `config_common.yml` this is suppressing it cause you're manually mapping the exception status in the tests:
```
        Symfony\Component\Serializer\Exception\ExceptionInterface: !php/const Symfony\Component\HttpFoundation\Response::HTTP_BAD_REQUEST
```

If you remove that line it correctly fails.

If you remove all the mappings  it correctly has status 400 again.

**Comment by soyuka** [comment]
Indeed configuration is not merged, you need to add back the defaults. We should add this to the documentation.

**Comment by KDederichs** [comment]
Yeah that would be useful, along with what the default actually is.
Or maybe just merge it as a new feature?
